ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/akkastreams/dispatcher/Dispatcher.scala

```diff
@@ -43,10 +43,15 @@ trait Dispatcher[Index] {
     */
   def shutdown(): Future[Unit]
 
-  /** Triggers shutdown of the Dispatcher by eagerly failing
-    * all outstanding stream subscriptions with the given throwable.
+  /** Triggers shutdown of the Dispatcher by eagerly failing all outstanding stream subscriptions with a throwable.
+    *
+    * @param throwableBuilder Create a new throwable.
+    *                     It is important to create a new throwable for each failed subscription
+    *                     since the throwable closing the streams or parts of it can be mutable.
+    *                     (e.g. the [[io.grpc.Metadata]] provided as part of [[io.grpc.StatusRuntimeException]]
+    *                     is mutated in the gRPC layer)
     */
-  def cancel(throwable: Throwable): Future[Unit]
+  def cancel(throwableBuilder: () => Throwable): Future[Unit]
```

Comment:
This is the only code path that can terminate a ledger subscription, right?

The approach cannot be easily transferred to Canton, because we may terminate subscriptions through code paths other than Dispatcher.cancel.

For Canton, I have the following approach in mind:
- Make sure to have an interceptor at every GRPC api.
- Whenever the interceptor sees a StatusRuntimeException/StatusException, it makes a defensive copy before passing the exception to GRPC.

That way, GRPC can never modify a Metadata that is being used elsewhere.

What are your thoughts on our approach?

Reply:
I was thinking around the same idea as @matthiasS-da 

An interceptor can actually see very custom exception, not necessarily `StatusRuntimeException`/`StatusException`, which is then translated to a proper gRPC status.

Reply:
> This is the only code path that can terminate a ledger subscription, right?

No. Ledger subscriptions can also be terminated in the streaming services machinery on close(). 

 > The approach cannot be easily transferred to Canton, because we may terminate subscriptions through code paths other than Dispatcher.cancel.

For the Daml SDK, I've chosen the approach of cleansing the code of reused throwables. You're saying that this is not an option for Canton (i.e. using unique StatusRuntimeExceptions for each closing stream)?

> Whenever the interceptor sees a StatusRuntimeException/StatusException, it makes a defensive copy before passing the exception to GRPC.

We already have an `ErrorInterceptor` in the Ledger API, so we could be defensive there and copy the Metadata but I avoided this approach since it's hacky and (e.g. `Metadata` has no copy method so we would need to do it ourselves). However, if you say that Canton needs to use this approach anyway, I'd rather use it in SDK as well.

Reply:
> For the Daml SDK, I've chosen the approach of cleansing the code of reused throwables. You're saying that this is not an option for Canton (i.e. using unique StatusRuntimeExceptions for each closing stream)?

Yes. In Canton, it would be hard to establish a guideline "whenever I fail a StreamObserver" I have to use a unique exception. So an interceptor would be more robust.

That being said, I haven't checked if an interceptor would be sufficient. I wonder if there are places other than failing a StreamObserver where we could hit this problem? 
Could be that we also may hit this (in Canton) if we serialize a StatusRuntimeException. So we may need even more than an interceptor.

Reply:
So, what would be the proper course of action here? Should we merge like this for Daml SDK and backport a metadata-copy solution when Canton implements it?

As a reference, here is the alternative for the metadata copy mitigation: https://github.com/digital-asset/daml/pull/15130/commits/56aa311a01b988b91ccb11dce7cd18912b4e93ee. I'm uneasy about it since it's working with io.grpc internal logic.

Reply:
CC @nmarton-da @matthiasS-da @mziolekda @skisel-da

Reply:
@tudor-da 
In my view, the most efficient path to a resolution is, if you choose a solution that works for the entire stack and push it through the entire stack. It takes you more time, because you have to touch the Canton repo and maybe even other components/repos that have GRPC servers. But it is still faster overall:
- Only you have to make yourself familiar with the Grpc issue that is causing all this.
- Only you have to invent a solution.
- We don't have to repeat the investigation when we hit this in other parts of the stack.
- You implement a minimal set of solution (as opposed to one solution per person that is working on this).

Concerning the metadata copy mitigation: I share your concern that we shouldn't use internal grpc api in general. However, when it comes to work around library bugs, I think it is ok to use internal api (or even reflection).

Reply:
We need to separate the very eminent problem with the db exceptions derailing the grpc streams that we have already observed and are making the 2.4.x releases broken from a broader issue of handling exception in the context of gRPC communication. The latter problem has been with us since always and we can spend more time working on a good and comprehensive solution.
Therefore I suggest we merge current solution as is and follow up with a new issue and a PR to deal with the general concern.

Reply:
The low level cloning approach what @matthiasS-da suggested is quite appealing, but the fact that it belongs to an internal API, I find very uncomfortable. Also if we would decide to go down this road, I would suggest following gRPC usage guidlines in docs, and contact them directly. Maybe we would learn something...if they answer.
I'd like to point out that gRPC java-docs clearly stating that Metadata is not thread safe, which for me means we were using the API incorrectly as opposed to facing some library bug here. (I am not saying I like, or we should like this API decision of the implementors)

Also looking under the hood it seems to me the Metadata's internal structure supports both serialized and unserialized storage, and if we use @tudor-da 's approach for cloning, we will used serialized (and based on the Exception stack traces I assume the original Metadata's are with the LazyValue-s). This also means from that point onwards as the gRPC internal code is working it's way with the Metadata it will hit different code-paths in Metadata as it would normally, which could open up another class of problems potentially. Documentation in Metadata states clearly: serialization intended to be used for transport purposes only.

On the other hand it seems to me that this problem is not typical: we are only hitting it if we stash somewhere pregenerated errors, or we want to send the same error to multiple places. Also Tudor only found 2 such places in the sdk AFAICS, so can it be that we are already in a nice place?
It seems to me the API really requires here distinct Metadata instances, so distinct errors/exceptions. We might consider simply respecting that, and staying on the yellow brick road?
Probably the same approach what we are following in here (pointwise fixes, plus general error detection) can be propagated to canton as well. WDYT @matthiasS-da ?

PS: @tudor-da you could use the InternalMetadata to access the serialize() without doing the ugliness of introducing io.grpc packages in our stack. They kindly opened it up (my guess would be not for us, though :) )

Reply:
> PS: @tudor-da you could use the InternalMetadata to access the serialize() without doing the ugliness of introducing io.grpc packages in our stack. They kindly opened it up (my guess would be not for us, though :) )

`InternalMetadata` has a bold disclaimer about not using it without contacting `io.grpc` team. 

@nmarton-da I agree with your thoughts as well. The grpc API design is not ideal and none of the proposed solutions are ideal. Probably the silver bullet here would be to use a Scala-compatible library for the gRPC layer (e.g. `akka-grpc`) as proposed by @skisel-da. 

But I'd like to park this discussion for now since we need a fix for the most pressing of the issues: closing the streams on Indexer restart and service teardown, which are likely to be more visible than the general problem.

Unless you have a strong opinion against it, please approve @nmarton-da @matthiasS-da .

Reply:
@tudor-da 
> Unless you have a strong opinion against it, please approve so we can deliver the pointwise fixes @nmarton-da @matthiasS-da .

Since you already have an approval, I won't take the time to go through this PR. (One approval is sufficient for merging, right?)

Having a point-wise fix is nice.

I have created this ticket to solve the wider problem, i.e., StatusRuntimeException not being thread-safe:
https://github.com/DACH-NY/canton/issues/10394

@mziolekda 
> We need to separate the very eminent problem with the db exceptions derailing the grpc streams that we have already observed and are making the 2.4.x releases broken from a broader issue of handling exception in the context of gRPC communication.

I expect that 2.3.x releases are also affected, as this problem was already observed in March 2022. Please make sure to backport your point-wise fix to the 2.3 release line.

Reply:
> ... and only manifesting itself as non-graceful shutdown of streams.

I disagree with this statement about the impact of the issue. This is not only about "non-graceful shutdown".

When we have hit this, the application became unavailable and a manual intervention was required (or would have been required) to get the application back into a healthy state. Automatic resubscription / failover did not work, effectively resulting in extended downtime. Several days have already been invested into investigating this.

Some high-profile clients are on 2.3.x. So I think, they should also get the fix.

Reply:
Sorry then, maybe I was making a wrong assumption about how this was manifesting in 2.3.x line based on my findings.

Reply:
Ok, thanks for explaining! Given that I agree the severity is lower for 2.3.x than for 2.4.x. Thanks for backporting it anyway! It's probably better to not take the risk.